src/main/java/picard/sam/util/SamComparison.java

```diff
@@ -1,67 +1,153 @@
+         * iterator (i.e. left file), for each duplicate set we create a set of duplicate reads that the representative read could "swap" with.  On the pass through the second iterator, for each
+         * duplicate read, we look for the reads it is allowed to swap with in the duplicate set from the first iterator.  If the allowed swap can also be performed in this duplicate set,
+         * we do not count the read or the read it could be swapped with as having mismatched duplicates.  Note however, that once a swap is allowed, the reads swapped cannot be used for any
+         * further swaps.
+         */
+        final HashMap<String, Set<String>> allowedSwapsMap = new HashMap<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetLeftIterator)) {
+            // do not want to redo duplicate marking, want to keep marks assigned in input files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            if (records.size() > 1) {
+                final Set<String> duplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                        .collect(Collectors.toSet());
+                records.stream().filter(r -> !r.getDuplicateReadFlag())
+                        .forEach(r -> allowedSwapsMap.put(r.getReadName(), duplicateRecordNames));
+            }
+        }
+        // set of fragments which have been "saved", i.e. will not be counted as having mismatched duplicate marks.
+        final HashSet<String> savedNames = new HashSet<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetRightIterator)) {
+            // do not want to redo duplicate marking, want to keep marking assigned in files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            if (records.size() > 1) {
+                final List<String> swappableDuplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                        .filter(allowedSwapsMap::containsKey).collect(Collectors.toList());
+                records.stream().filter(r -> !r.getDuplicateReadFlag()).map(SAMRecord::getReadName).filter(nRep -> !savedNames.contains(nRep))
+                    .forEach(nRep ->
+                            swappableDuplicateRecordNames.stream().filter(nDup -> !savedNames.contains(nDup) &&
+                                 allowedSwapsMap.get(nDup).contains(nRep))
+                                    .findFirst().ifPresent(nDup -> savedNames.addAll(Arrays.asList(nDup, nRep)))
+                    );
+            }
+            // count reads which are not saved
+            comparisonMetric.duplicateMarkingsDiffer += records.stream().map(SAMRecord::getReadName).filter(n -> !savedNames.contains(n)).count();
```

Comment:
I still find this code block to be pretty opaque - I made these changes in order to follow it - I'll leave it up to you to decide if you want to accept this suggested block if you think its clearer. Its just formatting, name changes and comments (I think!):

 ```suggestion
        final HashMap<String, Set<String>> swapTargetsForLeftReps = new HashMap<>();
        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetLeftIterator)) {
            // do not want to redo duplicate marking, want to keep marks assigned in input files
            final List<SAMRecord> allRecords = duplicateSet.getRecords(false);
            if (allRecords.size() > 1) {
                // for this set, get a list of non-rep reads that can potentially swap with the rep read
                final Set<String> nonRepReads = allRecords.stream()
                        .filter(SAMRecord::getDuplicateReadFlag)
                        .map(SAMRecord::getReadName)
                        .collect(Collectors.toSet());
                allRecords.stream().filter(r -> !r.getDuplicateReadFlag())
                        .forEach(leftRep -> swapTargetsForLeftReps.put(leftRep.getReadName(), nonRepReads));
            }
        }
        // set of fragments which have been "matched", i.e. will not be counted as having mismatched duplicate marks.
        final HashSet<String> matchedSwaps = new HashSet<>();
        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetRightIterator)) {
            // do not want to redo duplicate marking, want to keep marking assigned in files
            final List<SAMRecord> records = duplicateSet.getRecords(false);
            if (records.size() > 1) {
                // create a list of non-rep reads for this set that were also rep-reads in the left file
                final List<String> nonRepSwapCandidates = records.stream()
                        .filter(SAMRecord::getDuplicateReadFlag)
                        .map(SAMRecord::getReadName)
                        .filter(swapTargetsForLeftReps::containsKey)
                        .collect(Collectors.toList());
                // now try to match up the rep reads in this set with a non-rep swap candidate
                records.stream().filter(r -> !r.getDuplicateReadFlag()).map(SAMRecord::getReadName)
                        .filter(repRead -> !matchedSwaps.contains(repRead))
                        .forEach(unMatchedRep ->
                            nonRepSwapCandidates.stream()
                                    .filter(nonRep -> !matchedSwaps.contains(nonRep) && swapTargetsForLeftReps.get(nonRep).contains(unMatchedRep))
                                    .findFirst()
                                    .ifPresent(matchedRep -> matchedSwaps.addAll(Arrays.asList(matchedRep, unMatchedRep)))
                    );
            }
            // count reads which differ, but for which no swap match was found
            comparisonMetric.duplicateMarkingsDiffer += records.stream().filter(n -> !matchedSwaps.contains(n.getReadName())).count();
```

Reply:
Thanks for the clearer code!